### PR TITLE
Add logs to UI.

### PR DESCRIPTION
**Pull request by BramGruneir**
For now they just display (and only from the local node).  But it's a start.

screen shot 2015-06-11 at 23 48 30

### CODE REVIEW

**Comment by tamird** [comment]
Needs a rebase because of #1353

**Comment by BramGruneir** [comment]
Rebased and got some basic formatting for the log messages in.
This also includes a small fix for the Makefile as well.

screen shot 2015-06-12 at 13 24 17

File: resource/Makefile [hunk 1]
```diff
@@ -42,7 +42,7 @@ debug: $(GOBINDATA_DEBUG_TARGET)
 	rm -f $(GOBINDATA_TARGET)
 
 $(TYPESCRIPT_TARGET): $(shell find ui/ts)
-	tsc -p $^
+	tsc -p ui/ts
```
Comment (mrtracy) [inline resource/Makefile]: Why this change?

Comment (BramGruneir) [inline resource/Makefile]: the -p needs to point to a directory, the $^ is a list of ts files.

File: resource/ui/ts/util/convert.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
+		 */
+		export function MilliToNanos(millis: number): number {
+			return millis * 1.0e6;
+		}
+
+		/**
+		 * TimestampToDate converts a timestamp (nanos) into a Date.
+		 */
+		export function TimestampToDate(timestamp: number): Date {
+			return new Date(timestamp / 1.0e6);
```
Comment (mrtracy) [inline resource/ui/ts/util/convert.ts]: There's little indication other than the comment that the timestamp is in nanoseconds.

It's well known in Javascript that Date() accepts a millisecond unix timestamp.  Therefore, I think what you really want is `NanosToMilli`, and then in places where you are already using `TimestampToDate`, you really want to use `new Date(Convert.NanosToMilli(timestamp))`

File: resource/ui/ts/util/format.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     * Formatter contains common code for converting numbers and dates to human
+     * readable formats.
+     */
+    export module Format {
+        /**
+         * DateFromDate formats a Date object into a human readable date
+         * string.
+         */
+        var _datetimeFormatter = d3.time.format("%Y-%m-%d %H:%M:%S");
+        export function DateFromDate(datetime: Date): string {
```
Comment (mrtracy) [inline resource/ui/ts/util/format.ts]: This should just be Format.Date.  Everything in this package is going to output a string.

File: resource/ui/ts/util/format.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+         * SeverityFromNumber formats a numerical severity into its string
+         * representation.
+         */
+        var _severities = {
+            0 : "INFO",
+            1 : "WARNING",
+            2 : "ERROR",
+            3 : "FATAL"
+        };
+        export function SeverityFromNumber(severity: number): string {
```
Comment (mrtracy) [inline resource/ui/ts/util/format.ts]: Again, this should just be Format.Severity.  Also, consider adding Severity as an Enum in the proto library.

File: resource/ui/ts/models/proto.ts [hunk 5]
```diff
@@ -243,9 +243,46 @@ module Models {
...
+            json: string;
+        }
+
+        /**
+         * LogEntry represents a cockroach structured log entry.
+         *
+         * Source message = "LogEntry"
+         */
+        export interface LogEntry {
+            severity: number;
```
Comment (mrtracy) [inline resource/ui/ts/models/proto.ts]: This can probably be a typescript enum.

File: resource/ui/ts/models/log.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            private static _EntryRow(entry: Proto.LogEntry, count: number): _mithril.MithrilVirtualElement {
+                var dstyle, countStyle: string;
+                if (count % 2 == 0) {
+                    countStyle = tdStyleEvenFirst;
+                    dstyle = tdStyleEven;
+                } else {
+                    countStyle = tdStyleOddFirst;
+                    dstyle = tdStyleOdd;
+                }
+                return m("tr", [
```
Comment (mrtracy) [inline resource/ui/ts/models/log.ts]: This is View code, not Model code.  This was done on the Nodes and Status pages as well - the models contain rendering code.

What should really be done is to create a component that renders tabular data.  In the case of Node and Status pages tables were not appropriate, but in the case of logs there are multiple data columns, so I can buy a justification for using tables.

File: resource/ui/ts/models/log.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                    url += "entTime=" + this.endTime().toString() + "&";
+                }
+                if (this.max() != null) {
+                    url += "max=" + this.max().toString() + "&";
+                }
+                return url;
+            }
+
+            private _data = new Utils.QueryCache((): promise<Proto.LogEntry[]> => {
+                return m.request({ url: this._url(), method: "GET", extract: nonJsonErrors })
```
Comment (mrtracy) [inline resource/ui/ts/models/log.ts]: This has been factored into the Utils.Http package.  `Utils.Http.Get(this._url)` will work.

File: resource/ui/ts/models/log.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        var tdStyleOddFirst = "font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:#ccc;color:#333;background-color:#efefef;text-align:center";
+        var tdStyleOdd = "font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:#ccc;color:#333;background-color:#f9f9f9;text-align:center";
+        var tdStyleEvenFirst = "font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:#ccc;color:#333;background-color:#efefef;text-align:center";
+        var tdStyleEven = "font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:#ccc;color:#333;background-color:#fff;text-align:center";
+
+        export class Entries {
```
Comment (mrtracy) [inline resource/ui/ts/models/log.ts]: You should just extend this entire class from QueryCache.

```
export class Entries extends QueryCache<Proto.LogEntry[]> {
  constructor() {
    super(() => {
      // run your query here. 
    });
  }
}
```

Then you don't need the Refresh() or Entries() methods, you could just have results().

**Comment by petermattis** [comment]
The thread field is almost completely useless in Go logs. Goroutine id
would be useful, but we don't have access to it. I suggest just dropping
the thread column.

File: resource/ui/ts/models/log.ts (hunk 7 above)
Comment (mrtracy) [inline resource/ui/ts/models/log.ts]: I just realized I haven't checked it in yet.  It's out for PR now.

File: resource/ui/ts/util/convert.ts (hunk 2 above)
Comment (BramGruneir) [inline resource/ui/ts/util/convert.ts]: Done.

File: resource/ui/ts/util/format.ts (hunk 3 above)
Comment (BramGruneir) [inline resource/ui/ts/util/format.ts]: Good call.  Done.

File: resource/ui/ts/util/format.ts (hunk 4 above)
Comment (BramGruneir) [inline resource/ui/ts/util/format.ts]: I don't think it should be in the Proto Library unless it is a direct copy of a proto.  I'll change it to a enum, but we should find a good place for other constants.

File: resource/ui/ts/models/proto.ts (hunk 5 above)
Comment (BramGruneir) [inline resource/ui/ts/models/proto.ts]: Yes, but I don't know if this is the correct place for it.  I think this file should be purely typings direct from the cockroach protos only.

File: resource/ui/ts/models/log.ts (hunk 6 above)
Comment (BramGruneir) [inline resource/ui/ts/models/log.ts]: I'll move this code into the view.  And yay tables!
I'm fine turning these into some nice css based tables instead, I'm just not going to spend the time working on that yet.

File: resource/ui/ts/models/log.ts (hunk 7 above)
Comment (BramGruneir) [inline resource/ui/ts/models/log.ts]: Ah, good.  We can update in a further pr.

**Comment by BramGruneir** [comment]
I've dropped the tread column.

File: resource/ui/ts/models/log.ts (hunk 8 above)
Comment (BramGruneir) [inline resource/ui/ts/models/log.ts]: Awesome.  Done.

Comment (BramGruneir) [inline resource/ui/ts/models/log.ts]: Scratch that.  I can't figure out how to get this to work.  Since it's in the constructor, I can't get access to **this** in the super call, which I need for the **this._url()**.

**Comment by mrtracy** [comment]
LGTM
